File: go/numbers/alarm_clock.go
package numbers

import "fmt"

// Alarm Clock — schedule and trigger time-based alerts.

// AlarmEntry holds a single alarm's time and label.
type AlarmEntry struct {
	Hour   int
	Minute int
	Second int
	Label  string
}

// AlarmClock manages a collection of time-based alarms.
type AlarmClock struct {
	alarms []AlarmEntry
}

// NewAlarmClock creates an empty alarm clock.
func NewAlarmClock() *AlarmClock {
	return &AlarmClock{}
}

// SetAlarm schedules an alarm for the given time with an optional label.
func (ac *AlarmClock) SetAlarm(hour, minute, second int, label string) error {
	if hour < 0 || hour > 23 || minute < 0 || minute > 59 || second < 0 || second > 59 {
		return fmt.Errorf("invalid time: %02d:%02d:%02d", hour, minute, second)
	}
	ac.alarms = append(ac.alarms, AlarmEntry{
		Hour: hour, Minute: minute, Second: second, Label: label,
	})
	return nil
}

// Check compares all pending alarms against the given time (h, m, s).
// Returns the list of triggered alarms and removes them from pending.
func (ac *AlarmClock) Check(nowH, nowM, nowS int) []AlarmEntry {
	nowSecs := nowH*3600 + nowM*60 + nowS
	var triggered []AlarmEntry
	var remaining []AlarmEntry

	for _, a := range ac.alarms {
		alarmSecs := a.Hour*3600 + a.Minute*60 + a.Second
		if nowSecs >= alarmSecs {
			triggered = append(triggered, a)
		} else {
			remaining = append(remaining, a)
		}
	}
	ac.alarms = remaining
	return triggered
}

// Pending returns all alarms that have not yet triggered.
func (ac *AlarmClock) Pending() []AlarmEntry {
	result := make([]AlarmEntry, len(ac.alarms))
	copy(result, ac.alarms)
	return result
}

// Cancel removes the first pending alarm with the given label.
// Returns true if an alarm was removed, false otherwise.
func (ac *AlarmClock) Cancel(label string) bool {
	for i, a := range ac.alarms {
		if a.Label == label {
			ac.alarms = append(ac.alarms[:i], ac.alarms[i+1:]...)
			return true
		}
	}
	return false
}

// Demo usage (in a main package):
//
//	func main() {
//		clock := numbers.NewAlarmClock()
//		clock.SetAlarm(0, 0, 0, "midnight")
//		clock.SetAlarm(23, 59, 59, "end-of-day")
//		clock.SetAlarm(12, 0, 0, "noon")
//
//		fmt.Printf("Pending: %d\n", len(clock.Pending()))
//		clock.Cancel("noon")
//		triggered := clock.Check(12, 0, 0)
//		fmt.Printf("Triggered: %d\n", len(triggered))
//	}

File: rust/numbers/alarm_clock.rs
//! Alarm Clock — schedule and trigger time-based alerts.

/// A single alarm entry.
#[derive(Debug, Clone, PartialEq)]
pub struct Alarm {
    pub hour: u8,
    pub minute: u8,
    pub second: u8,
    pub label: String,
}

/// Manages a collection of time-based alarms.
///
/// The `check` method takes an explicit "now" tuple `(h, m, s)` so that
/// alarm logic is deterministic and testable without real clocks.
pub struct AlarmClock {
    alarms: Vec<Alarm>,
}

impl AlarmClock {
    /// Create an empty alarm clock.
    pub fn new() -> Self {
        AlarmClock { alarms: Vec::new() }
    }

    /// Schedule an alarm for the given time with an optional label.
    pub fn set_alarm(&mut self, hour: u8, minute: u8, second: u8, label: &str) {
        assert!(hour < 24 && minute < 60 && second < 60,
                "invalid time: {:02}:{:02}:{:02}", hour, minute, second);
        self.alarms.push(Alarm {
            hour,
            minute,
            second,
            label: label.to_string(),
        });
    }

    /// Check all pending alarms against the given time `(h, m, s)`.
    ///
    /// Returns the list of alarms that have triggered (time <= now).
    /// Triggered alarms are removed from the pending list.
    pub fn check(&mut self, now: (u8, u8, u8)) -> Vec<Alarm> {
        let now_secs = now.0 as u32 * 3600 + now.1 as u32 * 60 + now.2 as u32;
        let mut triggered = Vec::new();
        let mut remaining = Vec::new();

        for alarm in self.alarms.drain(..) {
            let alarm_secs = alarm.hour as u32 * 3600
                + alarm.minute as u32 * 60
                + alarm.second as u32;
            if now_secs >= alarm_secs {
                triggered.push(alarm);
            } else {
                remaining.push(alarm);
            }
        }
        self.alarms = remaining;
        triggered
    }

    /// Return a slice of all pending (not yet triggered) alarms.
    pub fn pending(&self) -> &[Alarm] {
        &self.alarms
    }

    /// Cancel the first pending alarm with the given label.
    ///
    /// Returns `true` if an alarm was removed, `false` otherwise.
    pub fn cancel(&mut self, label: &str) -> bool {
        if let Some(pos) = self.alarms.iter().position(|a| a.label == label) {
            self.alarms.remove(pos);
            true
        } else {
            false
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn set_and_pending() {
        let mut clock = AlarmClock::new();
        clock.set_alarm(7, 0, 0, "morning");
        clock.set_alarm(12, 0, 0, "noon");
        assert_eq!(clock.pending().len(), 2);
    }

    #[test]
    fn check_triggers_past_alarms() {
        let mut clock = AlarmClock::new();
        clock.set_alarm(6, 0, 0, "early");
        clock.set_alarm(23, 59, 59, "late");
        let triggered = clock.check((12, 0, 0));
        assert_eq!(triggered.len(), 1);
        assert_eq!(triggered[0].label, "early");
        assert_eq!(clock.pending().len(), 1);
    }

    #[test]
    fn check_triggers_exact_time() {
        let mut clock = AlarmClock::new();
        clock.set_alarm(8, 30, 0, "standup");
        let triggered = clock.check((8, 30, 0));
        assert_eq!(triggered.len(), 1);
        assert_eq!(triggered[0].label, "standup");
    }

    #[test]
    fn cancel_existing() {
        let mut clock = AlarmClock::new();
        clock.set_alarm(7, 0, 0, "morning");
        assert!(clock.cancel("morning"));
        assert_eq!(clock.pending().len(), 0);
    }

    #[test]
    fn cancel_nonexistent() {
        let mut clock = AlarmClock::new();
        clock.set_alarm(7, 0, 0, "morning");
        assert!(!clock.cancel("evening"));
        assert_eq!(clock.pending().len(), 1);
    }

    #[test]
    fn check_removes_triggered() {
        let mut clock = AlarmClock::new();
        clock.set_alarm(0, 0, 0, "midnight");
        let first = clock.check((12, 0, 0));
        assert_eq!(first.len(), 1);
        let second = clock.check((12, 0, 0));
        assert_eq!(second.len(), 0);
    }

    #[test]
    #[should_panic(expected = "invalid time")]
    fn invalid_hour() {
        let mut clock = AlarmClock::new();
        clock.set_alarm(25, 0, 0, "bad");
    }
}
